### ISSUE TITLE
More robust tests required for gradient checkpointing

### ISSUE THREAD

**Issue by anferico** [huggingface/transformers#32063]
As flagged in [this comment](https://github.com/huggingface/transformers/issues/31945#issuecomment-2233766500) from #31945, the [current test suite](https://github.com/huggingface/transformers/blob/72fb02c47dbbe1999ae105319f24631cad6e2e00/tests/test_modeling_common.py#L777) for gradient checkpointing may fail to spot subtle bugs. For example, the bug flagged in #31028 was not detected by any test, because the success criterion is simply that gradients are `not None` for each parameter that has `requires_grad=True`.

Proposal: every function whose behavior is affected by whether gradient checkpointing is enabled or not, such as a model's `forward()` method, should be tested twice: once with GC enabled and once with GC disabled, making sure that the outputs are exactly the same given certain inputs and configuration parameters.

**Comment by amyeroberts** [comment]
cc @ydshieh

**Comment by ydshieh** [comment]
Would you like to contribute @anferico ?

**Comment by anferico** [comment]
@ydshieh sure, though this will likely take some time and require support
